### ISSUE TITLE
Auto-detection of compilers with multiple names

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#518]
Recently, I discovered that certain PGI compilers (`pgf77` and `pgf90`) were unable to build HDF4, while newer ones (`pgfortran`) had no trouble. I submitted PR #493 to solve this issue. Now, if your compilers.yaml contains something like:

``` yaml
pgi@15.10-0:
      cc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgcc
      cxx: /soft/pgi/15.10/linux86-64/15.10/bin/pgc++
      f77: /soft/pgi/15.10/linux86-64/15.10/bin/pgfortran
      fc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgfortran
```

HDF4 will build for you. However, when I run:

``` bash
spack compiler add /soft/pgi/15.10/linux86-64/15.10/bin
```

Spack still finds the older compilers first:

``` yaml
pgi@15.10-0:
      cc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgcc
      cxx: /soft/pgi/15.10/linux86-64/15.10/bin/pgCC
      f77: /soft/pgi/15.10/linux86-64/15.10/bin/pgf77
      fc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgf90
```

The [PGI Installation Guide](https://www.pgroup.com/doc/pgiinstall.pdf) recommends using the newer compilers, although the older ones are kept around for backwards compatibility. Although I could simply remove `pgCC`, `pgf77`, and `pgf90` from those lists, it would cause problems for anyone using older installations of PGI, which don't come with the newer compilers.

I propose we change the way that `spack compiler add` finds compilers. My suggestion would be to choose the first match it finds in `cxx_names`, `f77_names`, and `f90_names`. I believe the relevant code is in the `find` method of `lib/spack/spack/compiler.py`. Things start to get complicated since we also need to be able to find compilers with prefixes, suffixes, or version numbers, and much of the code is parallelized for efficiency.

If anyone would like to tackle this, and maybe wrap in a solution to any of the unresolved problems we've been seeing with Apple compilers, be my guest. Otherwise, I will hopefully get to it eventually. It doesn't significantly impact me since I can just `spack config edit compilers` and switch it over to the newer ones, but it might trip up a new Spack user.

**Comment by tgamblin** [comment]
I'm forgetting if the mac stuff is the same problem or not, but I believe the PGI issues are a matter of making the code in [`Compiler.find()`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/compiler.py#L207) respect the order of names specified in the compiler descriptions, e.g. [these lines in `pgi.py`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/compilers/pgi.py#L34).

I can't get to it right now but I don't think it would be a major change.  The code just needs to sort the discovered compilers by version *and* the name preference order.

**Comment by becker33** [comment]
@adamjstewart is this issue still live?

**Comment by adamjstewart** [comment]
Yes, last time I checked. Spack chooses the first match it finds, not the first match in `cxx_names` and friends.

**Comment by becker33** [comment]
Fixed by #17967
